### ISSUE TITLE
Wrapping text issue in Japanese | Korean | Chinese 

### ISSUE THREAD

**Issue by ApoorvaDheeraj** [cocos/cocos-engine#15681]
### Cocos Creator version

2.4.6( Same issue in 3.7.3 Also)

### System information

macOS, Android,iOS, Window

### Issue description

I’m trying to wrap Japanese | Chinese label but its not wrapping properly. I'm using 2.4.6 ver but same issue can be seen in Cocos 3.7.3 also. 

Like, for example below , In Chinese text, you can see one symbol move to new line rather than whole word . 
`So i'm trying to solve this by moving whole word to new line, not only one symbol or character.`
Chinese -
![Screenshot 2023-07-11 at 3 16 40 PM](https://github.com/cocos/cocos-engine/assets/13048633/137099ef-2b55-4663-bfa9-0faa55473910)

Japanese - `キング・オブ・\nキャッシュ` last character should not display in new line whole word should be shifted to new line
![Screenshot 2023-07-11 at 1 27 16 PM](https://github.com/cocos/cocos-engine/assets/13048633/ad9617ff-bac8-4bd0-86c8-ab567bc2758e)

In `text-utils` class file, I can see we have added a range of symbols, but I don't know how "isUnicodeCJK" is actually working because it's not wrapping for these languages.

> __CHINESE_REG: /^[\u4E00-\u9FFF\u3400-\u4DFF]+$/,
> __JAPANESE_REG: /[\u3000-\u303F]|[\u3040-\u309F]|[\u30A0-\u30FF]|[\uFF00-\uFFEF]|[\u4E00-\u9FAF]|[\u2605-\u2606]|[\u2190-\u2195]|\u203B/g,
> 
> __KOREAN_REG: /^[\u1100-\u11FF]|[\u3130-\u318F]|[\uA960-\uA97F]|[\uAC00-\uD7AF]|[\uD7B0-\uD7FF]+$/,
> 
> isUnicodeCJK: function(ch) {
> return this.__CHINESE_REG.test(ch) || this.__JAPANESE_REG.test(ch) || this.__KOREAN_REG.test(ch);
> }

### Steps to reproduce

- Just create a label and  set width 600 and text to - キング・オブ・キャッシュ
- リーダーボードの頂点を極めてキング・オブ・キャッシュになりましょう 
`- and check for correct text wrap`

![Screenshot 2023-07-11 at 2 55 18 PM](https://github.com/cocos/cocos-engine/assets/13048633/82ab855f-3af6-4e1a-a918-fbe222d35d97)

### Minimal reproduction project

Font File
[AvertaCY-Bold.ttf.zip](https://github.com/cocos/cocos-engine/files/12013997/AvertaCY-Bold.ttf.zip)

**Comment by ApoorvaDheeraj** [comment]
@pandamicro @LinYunMo I'm trying to solve it by updating regex for these lang in [text-utils.ts](https://github.com/cocos/cocos-engine/blob/v3.7.2/cocos/2d/utils/text-utils.ts)

`Could you please suggest me how can add unicodes for these lang in optimize way.`

 Thanks

**Comment by mrtzd** [comment]
This wrapping text issue is also a problem for Arabic & Persian languages.

**Comment by ApoorvaDheeraj** [comment]
@mrtzd For temp solution you can add Arabic unicode range in [text-utlis.ts](https://github.com/cocos/cocos-engine/blob/v3.7.2/cocos/2d/utils/text-utils.ts) class.
You can find unicode range from here - https://jrgraphix.net/r/Unicode/

**Comment by mrtzd** [comment]
> @mrtzd For temp solution you can add Arabic unicode range in [text-utlis.ts](https://github.com/cocos/cocos-engine/blob/v3.7.2/cocos/2d/utils/text-utils.ts) class. You can find unicode range from here - https://jrgraphix.net/r/Unicode/

I'm not familiar with text rendering, so I'm not really sure exactly how to do that. Do you mean that I should add a function similar to "isUnicodeCJK" but for Arabic?

Anyway, thank you for the hint. I'll probably have to spend some more time learning how all of this works.

**Comment by LinYunMo** [comment]
> I'm trying to solve it by updating regex for these lang in [text-utils.ts](https://github.com/cocos/cocos-engine/blob/v3.7.2/cocos/2d/utils/text-utils.ts)
> 
> `Could you please suggest me how can add unicodes for these lang in optimize way.`
> 
> Thanks

Sorry I'm late. @ApoorvaDheeraj 

**About the Chinese part：**

![image](https://github.com/cocos/cocos-engine/assets/42071583/57d498eb-be14-4782-865c-72699ad8a9ee)

In Chinese line breaks, there is no need for complete words that can't be split, the above result is correct.

In fact, for CJK, the requirements for line breaks are the same, there is no such thing as a letter that cannot be split, and unlike in English, a letter split does not affect the meaning of the sentence.

**About the Japanese section**

![image](https://user-images.githubusercontent.com/13048633/252644294-ad9617ff-bac8-4bd0-86c8-ab567bc2758e.png)

Only some of the letters in Japanese are not separable from each other.

For example, the effect of the above picture is incorrect, but the effect of the following picture is all correct :

![企业微信截图_16903663642671](https://github.com/cocos/cocos-engine/assets/42071583/61690ec8-43af-4aee-b372-3e455904cf51)

![企业微信截图_16903664602342](https://github.com/cocos/cocos-engine/assets/42071583/6a950fdd-d0c0-43fc-8b5d-ae3b056e8c72)

![企业微信截图_16903664861940](https://github.com/cocos/cocos-engine/assets/42071583/11e05a51-0a12-4382-8cc5-bb876b273ea0)

I need to find some more information about the exact combination rules, It has been observed that line breaks in css don't behave exactly correctly either.

It's safe to say that there is no such rule in Chinese, only the convention that punctuation marks should be kept out of the beginning of the line as much as possible. 

***Is it possible to use shrink mode to solve this? It seems that the size of the text in a project should be the same in all languages, otherwise there will be typographical problems, so is it possible to use the shrink mode to ensure that there will be no differences in line breaks by specifying the width and automatically calculating the font size?***

**Comment by LinYunMo** [comment]
> @mrtzd For temp solution you can add Arabic unicode range in [text-utlis.ts](https://github.com/cocos/cocos-engine/blob/v3.7.2/cocos/2d/utils/text-utils.ts) class. You can find unicode range from here - https://jrgraphix.net/r/Unicode/

This part of the code is just to separate out the CJK text, which has different line breaks due to its.
The wrapping rules are in the ```text-utils``` class file:
```
const WORD_REG = /([a-zA-Z0-9ÄÖÜäöüßéèçàùêâîôûа-яА-ЯЁё]+|\S)/;
// eslint-disable-next-line no-useless-escape
const SYMBOL_REG = /^[!,.:;'}\]%\?>、‘“》？。，！]/;
const LAST_WORD_REG = /([a-zA-Z0-9ÄÖÜäöüßéèçàùêâîôûаíìÍÌïÁÀáàÉÈÒÓòóŐőÙÚŰúűñÑæÆœŒÃÂãÔõěščřžýáíéóúůťďňĚŠČŘŽÁÍÉÓÚŤżźśóńłęćąŻŹŚÓŃŁĘĆĄ-яА-ЯЁёáàảạãăắằẳẵặâấầẩẫậéèẻẽẹêếềểễệiíìỉĩịóòỏõọôốồổỗộơớờởỡợúùủũụưứừửữựýỳỷỹỵđÁÀẢẠÃĂẮẰẲẴẶÂẤẦẨẪẬÉÈẺẼẸÊẾỀỂỄỆIÍÌỈĨỊÓÒỎÕỌÔỐỒỔỖỘƠỚỜỞỠỢÚÙỦŨỤƯỨỪỬỮỰÝỲỶỸỴĐ]+|\S)$/;
const LAST_ENGLISH_REG = /[a-zA-Z0-9ÄÖÜäöüßéèçàùêâîôûаíìÍÌïÁÀáàÉÈÒÓòóŐőÙÚŰúűñÑæÆœŒÃÂãÔõěščřžýáíéóúůťďňĚŠČŘŽÁÍÉÓÚŤżźśóńłęćąŻŹŚÓŃŁĘĆĄ-яА-ЯЁёáàảạãăắằẳẵặâấầẩẫậéèẻẽẹêếềểễệiíìỉĩịóòỏõọôốồổỗộơớờởỡợúùủũụưứừửữựýỳỷỹỵđÁÀẢẠÃĂẮẰẲẴẶÂẤẦẨẪẬÉÈẺẼẸÊẾỀỂỄỆIÍÌỈĨỊÓÒỎÕỌÔỐỒỔỖỘƠỚỜỞỠỢÚÙỦŨỤƯỨỪỬỮỰÝỲỶỸỴĐ]+$/;
const FIRST_ENGLISH_REG = /^[a-zA-Z0-9ÄÖÜäöüßéèçàùêâîôûаíìÍÌïÁÀáàÉÈÒÓòóŐőÙÚŰúűñÑæÆœŒÃÂãÔõěščřžýáíéóúůťďňĚŠČŘŽÁÍÉÓÚŤżźśóńłęćąŻŹŚÓŃŁĘĆĄ-яА-ЯЁёáàảạãăắằẳẵặâấầẩẫậéèẻẽẹêếềểễệiíìỉĩịóòỏõọôốồổỗộơớờởỡợúùủũụưứừửữựýỳỷỹỵđÁÀẢẠÃĂẮẰẲẴẶÂẤẦẨẪẬÉÈẺẼẸÊẾỀỂỄỆIÍÌỈĨỊÓÒỎÕỌÔỐỒỔỖỘƠỚỜỞỠỢÚÙỦŨỤƯỨỪỬỮỰÝỲỶỸỴĐ]/;
```

**Comment by ApoorvaDheeraj** [comment]
> Is it possible to use shrink mode to solve this?

@LinYunMo  No, because of some project related things we can't change Overflow method to Shrink. 

And As you said for **Japanese** [For example, the effect of the above picture is incorrect](https://github.com/cocos/cocos-engine/issues/15681#:~:text=For%20example%2C%20the%20effect%20of%20the%20above%20picture%20is%20incorrect%2C%20but%20the%20effect%20of%20the%20following%20picture%20is%20all%20correct) where last character moved to new line which is incorrect, Can we solve for these thing in CJK ?

And One more thing i want to know about performance - I want Support For **Hindi** and Unicode Range is **0900 — 097F** and i'm planning to append Unicode Range for this in regex pattern. 

So i just wanted to know adding more ranges affect the performance ? _IF Yes - Then how can i debug this thing._ 


> const LAST_WORD_REG = /([a-zA-Z0-9ÄÖÜäöüßéèçàùêâîôûаíìÍÌïÁÀáàÉÈÒÓòóŐőÙÚŰúűñÑæÆœŒÃÂãÔõěščřžýáíéóúůťďňĚŠČŘŽÁÍÉÓÚŤżźśóńłęćąŻŹŚÓŃŁĘĆĄ-яА-ЯЁёáàảạãăắằẳẵặâấầẩẫậéèẻẽẹêếềểễệiíìỉĩịóòỏõọôốồổỗộơớờởỡợúùủũụưứừửữựýỳỷỹỵđÁÀẢẠÃĂẮẰẲẴẶÂẤẦẨẪẬÉÈẺẼẸÊẾỀỂỄỆIÍÌỈĨỊÓÒỎÕỌÔỐỒỔỖỘƠỚỜỞỠỢÚÙỦŨỤƯỨỪỬỮỰÝỲỶỸỴĐ\u0900—\u097F]+|\S)$/;

You can find unicode ranges for Hindi [Here](https://jrgraphix.net/r/Unicode/0900-097F)

**Comment by ApoorvaDheeraj** [comment]
[Here](https://github.com/cocos/cocos-engine/issues/15681#:~:text=This%20part%20of%20the%20code%20is%20just%20to%20separate%20out%20the%20CJK%20text%2C%20which%20has%20different%20line%20breaks%20due%20to%20its.) you mentioned it's for separating CJK from other lang which have different line break rule. 

**So for CJK we don't check for any line breaks thing ?**

**Comment by LinYunMo** [comment]
> **So for CJK we don't check for any line breaks thing ?**

If there are any line breaks in the string of the label, it will be recognized and line breaks will be applied. Like type a direct line change in the String Box.
Other than that, only punctuation is guaranteed not to appear at the beginning of a line。

The above rule is true in all Chinese and Korean, but there are some consonants that cannot be split in Japanese。
If want to add a break rule, we need to add to the function ```fragmentText```

**Comment by LinYunMo** [comment]
> So i just wanted to know adding more ranges affect the performance ? _IF Yes - Then how can i debug this thing._

Yes, it will have an effect on performance.  If there are different rules, add a new ranges at function ```fragmentText```.

This may be able to solve the problem, but for a complete solution to the breakage problem, additional libraries would need to be introduced, which would cause performance degradation and a significant increase in package size, which is not a simple problem that we will try to solve in the future.

At this stage, try changing``` LAST_WORD_REG``` ```LAST_WORD_REG``` and function ```fragmentText``` to solve the problem.

**Comment by ApoorvaDheeraj** [comment]
@LinYunMo Are you working on  this test case [Only some of the letters in Japanese are not separable from each other](https://github.com/cocos/cocos-engine/issues/15681#:~:text=Only%20some%20of%20the%20letters%20in%20Japanese%20are%20not%20separable%20from%20each%20other.)

**Comment by LinYunMo** [comment]
Sorry, we don't have a schedule to optimize this issue at the moment
